src/ActivityPub/ActorHandle.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\ActivityPub;
+
+class ActorHandle
+{
+    public const HANDLE_PATTERN = '/^(?P<prefix>[@!])?(?P<user>[\w\-\.]+)@(?P<host>[\w\.\-]+)(?P<port>:[\d]+)?$/';
+
+    public function __construct(
+        public ?string $prefix = null,
+        public ?string $username = null,
+        public ?string $host = null,
+        public ?int $port = null,
+    ) {
+    }
+
+    public function __toString(): string
+    {
+        return $this->formatWithPrefix($this->prefix);
+    }
+
+    public static function parse(string $handle): ?static
+    {
+        if (preg_match(static::HANDLE_PATTERN, $handle, $match)) {
+            $new = new static(
+                $match['prefix'] ?? null,
+                $match['user'],
+                $match['host']
+            );
+            $new->setPort($match['port'] ?? null);
+
+            return $new;
+        }
+
+        return null;
+    }
+
+    public static function isHandle(string $handle)
+    {
+        if (preg_match(static::HANDLE_PATTERN, $handle, $matches)) {
+            return !empty($matches['user']) && !empty($matches['host']);
+        }
+
+        return false;
+    }
+
+    public function isValid(): bool
+    {
+        return static::isHandle((string) $this);
+    }
+
+    /** @return string port as string in the format ':9000' or empty string if it's null */
+    public function getPortString(): string
+    {
+        return !empty($this->port) ? ':'.$this->port : '';
+    }
+
+    /**
+     * @param int|string|null $port port as either plain int or string formatted like ':9000'
+     */
+    public function setPort(int|string|null $port)
+    {
+        if (\is_string($port)) {
+            $this->port = \intval(ltrim($port, ':'));
+        } else {
+            $this->port = $port;
+        }
+
+        return $this;
+    }
+
+    public function getDomain(): string
+    {
+        return $this->host.$this->getPortString();
+    }
+
+    public function setDomain(?string $domain)
+    {
+        $url = parse_url($domain);
+
+        $this->host = $url['host'] ?? null;
+        $this->port = $url['port'] ?? null;
+
```

Comment:
Is null a valid default? Or 443?

Reply:
default `null` is fine, it means that there's no port number after the host part i.e. the handle it holds is just `username@domain` and not `username@domain:443`

this class is first and foremost a string manipulation tool, it does not know or care that https is port 443 and that should be the default